### ISSUE TITLE
IR3 LoadersStacking Error

### ISSUE THREAD

**Issue by Kyra-Lee** [Deadlock989/IndustrialRevolution#402]
Looks like I have an error with the IR3 LoadersStacking Mod.  I have tested this with the base IR3 mod only and with a few of the addons like airships, and I get this same error each time.  I am 99% sure it only happens when I enable the IR3 LoadersStacking mod.  From what I can tell I have satisfied all of the mods dependencies.  Below is the error text.  Let me know if I should provide more info.

Error loading mods

Failed to load mods: Error in assignID: recipe-category with name 'unstacking' does not exist.
Source: character (character).

Mods to be disabled:
· IndustrialRevolution3 (3.1.20)
. IndustrialRevolution3LoadersStacking (1.0.9)
. IndustrialRevolution3 (3.1.20)

Factorio version - 1.1.104 (build 62119, win64)

Mods Loaded(and errored out)
![image](https://github.com/Deadlock989/IndustrialRevolution/assets/46364330/5b4f86e0-a670-4aad-a76a-a574c9ce6f67)

**Comment by Deadlock989** [comment]
This is very strange. Do you have any other mods installed?

**Comment by Kyra-Lee** [comment]
I did some more testing and it seems this might be related to the startup settings on one of the IR3 mods. Might still be the Loader one.  I did a complete re-install of Factorio and wiped out the mods folder so I could start from scratch.  I loaded each individual mod one at a time and never ran into errors.  From there I tried to load up a save game I had from Factorio version 1.1.94-0 with IR3 Version 3.1.12 and LoadersStacking version 1.0.5 and when asked to sync mods and startup settings I did not do that and just loaded the game. Everything worked fine.  I closed that game out and loaded it again, this time syncing mods and startup settings.  It didn't download any mods but still crashed with a similar error as above.  I think this is related to the startup settings specifically.

I do not know what startup settings I had selected before, its been a few months since I played on that save.  I loaded that save up without syncing startup settings, then created a new save.  I closed that game, turned off all of my mods, and restarted the client.  After that I loaded my new save up, synced mods and startup settings and everything worked.

I read through the patch notes for both main IR3 and the Loader mod from my old version to the newest and I didn't see anything that stood out to me as a breaking issue, but maybe you can see what happened.  For now this seems like a very edge case scenario with converting to newer versions from old ones.  I was able to work around my issue and I don't think it will be a problem in the future.  If you want to continue investigating feel free to let me know how I can provide more info.

Thanks,
KL

**Comment by Deadlock989** [comment]
I can't currently explain the face value interpretation of this error. I would be surprised if it was being triggered by start-up settings because neither of Loaders & Stacking's start-up settings have anything to do with the creation of recipe categories. Even if they did, I can't currently explain how the mod could get into that state. The mod unconditionally creates that recipe category in the data-updates stage, and then unconditionally gives it to the character entity in the data-final-fixes - it does not consult settings to do this. But the error is suggesting that the recipe category was not created. So there are two possibilities:

- Some other mod is deleting the recipe category (unlikely)
- The error is a false positive and is merely indicative of some other kind of more fundamental problem such as file corruption (I have dim memories of seeing weird and inappropriate assignID errors in the logs some time ago, but forget the conditions they arose under) 

I can't solve problems without being able to reproduce them reliably. If you wanted to provide a copy of that older saved game, I could try following your steps to see if it occurs. But my gut feeling is that this is some difficult-to-reproduce combination of issues arising from syncing between different game and mod versions, and/or the presence of malformed data somewhere.

**Comment by Kyra-Lee** [comment]
Here is that older save you requested.  I forgot that I did start a "new game" with the New Game+ mod by Bilka.  I didn't run into any issues at the time when I did that several months ago.
[IR3 Plus Save 1.zip](https://github.com/Deadlock989/IndustrialRevolution/files/14581723/IR3.Plus.Save.1.zip)

Older save before I made Used the New Game Plus mod.  This one loads just fine.
[IR3-Save2.zip](https://github.com/Deadlock989/IndustrialRevolution/files/14581752/IR3-Save2.zip)

I completely forgot about the New Game+ mod.  Hopefully this didn't end up being a goose chase.

Thanks,
KL

**Comment by Deadlock989** [comment]
Thank you for providing the saves.

I was able to reproduce the same error with the Save 1 file. You did have several other mods installed (AAI Containers, Alien Biomes + HR terrain, Text Plates + Even More Text Plates, FNEI, Larger Lamps, Light Overhaul, Max Rate Calculator, RSO).

Disabling random other mods one by one, I was able to get the save to load (without resetting mod settings) just once. Repeating this experiment did not work a second time.

However the save can be loaded immediately and reliably with all of the mods referenced in the save file if, at the first time you get the "Error loading mods" dialog and the warning about the recipe category, you tick the "Reset mod settings" button next to IndustrialRevolution3LoadersStacking and then click Restart.

I went back and checked the version of LoadersStacking your save references (1.0.4) with the current version (1.0.9). There have been no changes made to the mod's settings in that time. The mod doesn't take any new conditional actions based on settings either.

I can't explain this. It may be that the mod settings data in that Save 1 is somehow corrupted (I can Save 2 every time without issues) and this corruption is somehow producing a completely misleading error message. Or it may be some other issue with Factorio. But I am 90% confident it is not an issue produced by LoadersStacking itself.

Thanks for bringing it to my attention, if I get any more reports like this and any kind of sense emerges, I'll look into it again.
